conmon-rs/server/src/attach.rs

```diff
@@ -44,43 +49,82 @@ impl Attach {
+        fd: RawFd,
+        stdin: Sender<String>,
+        stdout: Sender<Vec<u8>>,
+        stderr: Sender<Vec<u8>>,
+    ) -> Result<()> {
         debug!("Start listening on attach socket");
-        let listener = unsafe { UnixListener::from_raw_fd(fd) };
-        for stream in listener.incoming() {
-            thread::spawn(|| Self::handle_client(stream?));
+        let listener = UnixListener::from_std(unsafe { net::UnixListener::from_raw_fd(fd) })?;
+        loop {
+            match listener.accept().await {
+                Ok((stream, _)) => {
+                    let stdin_tx = stdin.clone();
+                    let stdout_rx = stdout.subscribe();
+                    let stderr_rx = stderr.subscribe();
+                    task::spawn(async move {
+                        if let Err(e) =
+                            Self::handle_client(stream, stdin_tx, stdout_rx, stderr_rx).await
+                        {
+                            error!("Unable to handle attach client: {:#}", e)
+                        }
+                    });
+                }
+                Err(e) => error!("Unable to accept attach stream: {}", e),
+            }
         }
-        Ok(())
     }
 
-    fn handle_client(stream: UnixStream) -> Result<()> {
+    async fn handle_client(
+        stream: UnixStream,
+        stdin: Sender<String>,
+        mut stdout: Receiver<Vec<u8>>,
+        mut stderr: Receiver<Vec<u8>>,
+    ) -> Result<()> {
         debug!("Got new attach stream connection");
 
-        let mut reader = BufReader::new(&stream);
-        let mut writer = BufWriter::new(&stream);
-
-        const BUF_SIZE: usize = 8192;
-        let mut buf = vec![0; BUF_SIZE];
-
-        // TODO: implement me
-        buf[0] = 2;
-        writer.write_all(&buf)?;
-        buf[0] = 3;
-        writer.write_all(&buf)?;
+        enum Pipe {
+            Stdout = 2,
+            Stderr = 3,
+        }
 
+        let mut io = BufStream::new(stream);
         loop {
-            match reader.read(&mut buf) {
-                Ok(n) if n > 0 => {
-                    debug!("Read {} bytes", n);
-                }
-                Err(e) => error!("Unable to read from file descriptor: {}", e),
-                _ => {}
-            }
+            let mut line = String::new();
+
+            select! {
+                res = stdout.recv() => {
+                    let mut data = res?;
+                    debug!("Got {} byte stdout data", data.len());
+                    data.insert(0, Pipe::Stdout as u8);
+                    io.write_all(&data).await.context("write stdout")?;
+                },
+                res = stderr.recv() => {
+                    let mut data = res?;
+                    debug!("Got {} byte stderr data", data.len());
+                    data.insert(0, Pipe::Stderr as u8);
+                    io.write_all(&data).await.context("write stderr")?;
+                },
+                res = io.read_line(&mut line) => {
```

Comment:
Careful with `select!` and stateful futures - https://blog.yoshuawuyts.com/futures-concurrency-3/

(In response to reading that blog I did https://github.com/ostreedev/ostree-rs-ext/pull/239 )

In this case the simplest fix may be to use e.g. https://docs.rs/tokio/latest/tokio/io/struct.Lines.html - then pin/box that future, and then you can read from that stream alongside the channels.

Reply:
Ugh, we should probably do that for the whole project at once. Created https://github.com/containers/conmon-rs/issues/318 for that as follow-up.

Reply:
Another option I think is to e.g. `Box::pin` the `io.read_line` future - that moves its state outside.

In general it's worth contrasting with e.g. Go - there every goroutine has a stack, which means one doesn't have to think about this type of problem.  But OTOH, cancellation is totally different and much more manual, and async Rust's "stackless" coroutine approach can be a lot more efficient in hot paths.